### ISSUE TITLE
Brand new Italian CNS (ACe 2021) not working: C_Login failed: rv = CKR_USER_PIN_NOT_INITIALIZED (0x102)

### ISSUE THREAD

**Issue by nestukh** [OpenSC/OpenSC#2445]
Hello I have a brand new Italian CNS ('Tessera Sanitaria', TS-CNS), the provider is ACe 2021 (official driver [here](https://idistribution.it/bit4id-minilector/)). After rebuilding OpenSC from git cloning, on Linux (Debian Bullseye amd64) I'm still can't login via its PIN, I'm getting `C_Login failed: rv = CKR_USER_PIN_NOT_INITIALIZED (0x102)`, in a similar way as described in https://github.com/OpenSC/OpenSC/issues/1995. I've rebuilt CCID too.
My previous TS-CNS smartcard had no problem with the same setup, but the provider was OT2015.

Logs:
the command (OpenSC built in `$HOME/.bin/TS-CNS/static/`)
```
$HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /home/$USER/.bin/TS-CNS/static/lib/opensc-pkcs11.so
```
results in
```
Using slot 0 with a present token (0x0)
error: PKCS11 function C_Login failed: rv = CKR_USER_PIN_NOT_INITIALIZED (0x102)
Aborting.
```
the full logs of 
`OPENSC_DEBUG=3 $HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /home/$USER/.bin/TS-CNS/static/lib/opensc-pkcs11.so`
are here (some parts omitted for privacy): https://gist.github.com/nestukh/29ddd4e55f63bcb207e8f158d41a69d3

Using the official driver for this card (`/usr/lib/bit4id/libbit4opki.so`), the command
```
$HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /usr/lib/bit4id/libbit4opki.so
```
returns
```
Using slot 0 with a present token (0x0)
Logging in to "Carta Nazionale dei Servizi".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seems to be OK
Digests:
  all 4 digest functions seem to work
  SHA-1: OK
  SHA256: OK
  ERR: C_Digest() didn't return CKR_BUFFER_TOO_SMALL but CKR_GENERAL_ERROR (0x5)
  ERR: C_Digest() didn't return CKR_OK for a NULL output buffer, but CKR_OPERATION_NOT_INITIALIZED (0x91)
  ERR: digest operation ended prematurely
Signatures (currently only for RSA)
  testing key 0 (CNS1) 
  ERR: C_SignUpdate failed: CKR_FUNCTION_FAILED (0x6)
  ERR: C_Sign() didn't return CKR_BUFFER_TOO_SMALL but CKR_GENERAL_ERROR (0x5)
  ERR: C_Sign() didn't return CKR_OK for a NULL output buf, but CKR_OPERATION_NOT_INITIALIZED (0x91)
error: PKCS11 function C_Sign failed: rv = CKR_GENERAL_ERROR (0x5)
Aborting.
```
the full log with `OPENSC_DEBUG=3` doesn't differ (I don't know why)

**Comment by nestukh** [comment]
Addendum: I'm using this snipplet to install everything

```
### OpenSC
sudo apt-get install -y p11-kit libusb-1.0-0-dev pcscd pcsc-tools libpcsclite-dev libgempc430 libasedrive-usb libacsccid1 libccid  libpcsclite1 opensc libnss3-tools libnss3-dev libnss3 opensc opensc-pkcs11 libssl-dev zlib1g-dev libreadline-dev autoconf automake build-essential docbook-xsl xsltproc libtool pkg-config <&- # Network Security Service tools (Mozilla-nss-tools) included
sudo /etc/init.d/pcscd restart #  pcsc_scan 

cd
installpath="$HOME/.bin/TS-CNS"
rm -fRr "$installpath"
mkdir -p "$installpath/static/bin" "$installpath/static/lib"
cd "$installpath"
git clone https://github.com/OpenSC/OpenSC
cd OpenSC
./bootstrap
./configure --prefix="$installpath/static" --datadir="$installpath/static"
make
sudo make install # sudo for "libtool: install: /usr/bin/install -c .libs/pkcs15-tool /home/$USER/.bin/TS-CNS/static/bin/pkcs15-tool"   etc
cd

### CCID
cd
installpath="$HOME/.bin/TS-CNS"
rm -fRr "$installpath/CCID"
rm -f "$installpath/ccid-"*".tar.bz2"*
mkdir -p "$installpath/static/bin" "$installpath/static/lib"
cd "$installpath"
# git clone https://salsa.debian.org/rousseau/CCID
wget "https://ccid.apdu.fr/files/$(wget -O- -q https://ccid.apdu.fr/files/ | grep -A2 -i Latest | grep "href=\"ccid-" | sed "s/.tar.bz2\".*$/.tar.bz2/g;s/^.*href=\"//g")"
tar xvf "$(ls -1 | grep -i ccid | grep -i .tar.bz2)" -C $PWD
mv "$(ls -1 | grep -i ccid | grep -vi .tar.bz2)" CCID
cd CCID
./configure # --prefix="$installpath/static" --datadir="$installpath/static" --enable-usbdropdir="$installpath/static/usr/local/pcsc/drivers" --enable-usbdropdir="$installpath/static/usr/local/pcsc/drivers" --enable-ccidtwindir="$installpath/static/usr/local/pcsc/drivers/serial" --enable-serialconfdir="$installpath/static/usr/local/pcsc/drivers"
make
sudo make install
sudo /etc/init.d/pcscd restart

**Comment by Jakuje** [comment]
Reading the logs, the itacns driver is correctly tried:
```
P:90108; T:0x139944898844736 10:32:10.644 [opensc-pkcs11] card.c:341:sc_connect_card: trying driver 'itacns'
```
is correctly matched:
```
P:90108; T:0x139944898844736 10:32:10.645 [opensc-pkcs11] card.c:355:sc_connect_card: matched: Italian CNS
```
but then you got some failures to select files.

Later on, the driver fails to parse the certificates:
```
P:90108; T:0x139944898844736 10:32:11.365 [opensc-pkcs11] pkcs15-cert.c:93:parse_x509_cert: X.509 certificate not found: -1401 (Invalid ASN.1 object)
P:90108; T:0x139944898844736 10:32:11.365 [opensc-pkcs11] pkcs15-cert.c:403:sc_pkcs15_read_certificate: returning with: -1401 (Invalid ASN.1 object)
P:90108; T:0x139944898844736 10:32:11.366 [opensc-pkcs11] pkcs15-itacns.c:233:itacns_add_cert: Could not read X.509 certificate: -1401 (Invalid ASN.1 object)
```
but still it looks like it found one:
```
P:90108; T:0x139944898844736 10:32:11.423 [opensc-pkcs11] framework-pkcs15.c:901:pkcs15_create_pkcs11_objects: Found 1 certificate
```
It has problems to find pins, which would suggest the problems you have with login:
```
P:90108; T:0x139944898844736 10:32:11.422 [opensc-pkcs11] pkcs15.c:1745:sc_pkcs15_find_pin_by_flags: Find PIN flags:0x10, mask:0xD2, index:-1
P:90108; T:0x139944898844736 10:32:11.422 [opensc-pkcs11] pkcs15.c:1771:sc_pkcs15_find_pin_by_flags: returning with: -1407 (Requested object not found)
P:90108; T:0x139944898844736 10:32:11.422 [opensc-pkcs11] pkcs15.c:1744:sc_pkcs15_find_pin_by_flags: called
P:90108; T:0x139944898844736 10:32:11.422 [opensc-pkcs11] pkcs15.c:1745:sc_pkcs15_find_pin_by_flags: Find PIN flags:0x12, mask:0xD2, index:-1
P:90108; T:0x139944898844736 10:32:11.422 [opensc-pkcs11] pkcs15.c:1771:sc_pkcs15_find_pin_by_flags: returning with: -1407 (Requested object not found)
```
If it is a new card, there might be some needed for the driver adjustments based on the specifications, but it will require some more investigation.

**Comment by nestukh** [comment]
The smartcard is very brand new (I've received it by mail last week, and the PIN, PUK and some other code by email this week).
On raspberry pi OpenSC is extra useful, as the official driver is only amd64, so I'm available to any test you might need.

**Comment by frankmorgner** [comment]
@puccia @resoli , any ideas?

**Comment by resoli** [comment]
Hello, unfortunately it seems that 2021 CNS health cards are problematic, I had several reports from users (see https://github.com/OpenSC/OpenSC/issues/2370#issuecomment-972641235 for a "ST2021" card ). 
I assume that @nestukh has correctly activated the new card using its previous TS-CNS ...

**Comment by nestukh** [comment]
Yes it's activated and I can log in on various websites (Agenzia delle Entrate, INPS) via firefox, using the official driver and the PIN of my smarcard.
I was already aware of https://github.com/OpenSC/OpenSC/issues/2370 and the general issue that permeates this Country.

**Comment by resoli** [comment]
Ok, I see that even using the proprietary module thru opensc:
`$HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /usr/lib/bit4id/libbit4opki.so`
is problematic.

Recently I had another report from a user for 
`PKCS11 function C_Login failed: rv = CKR_USER_PIN_NOT_INITIALIZED (0x102)`
I have to ask him if the card is an ACe 2021 as well

**Comment by resoli** [comment]
>  I have to ask him if the card is an ACe 2021 as well

Confirmed, ACe 2021 issued from Valle d'Aosta region.

**Comment by resoli** [comment]
Maybe pkcs11-spy with the proprietary module + OPENSC_DEBUG=9 could help?

**Comment by nestukh** [comment]
If you mean something like this
```
PKCS11SPY=/usr/lib/bit4id/libbit4opki.so PKCS11SPY_OUTPUT=$HOME/pkcs11spy_log.txt OPENSC_DEBUG=9 $HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /home/$USER/.bin/TS-CNS/static/lib/pkcs11-spy.so
```
Here are the logs [`pkcs11spy_log.txt`](https://gist.github.com/nestukh/a449819f2eb4e2ddba8c73240694c1af) (privacy protected, if I'd erased too much data just tell me the lines you need)

**Comment by resoli** [comment]
Yes, it would be more useful with the correspondent standard output with debug level 9 (as usual be aware of sharing a curated version).

**Comment by puccia** [comment]
I haven't worked on any smartcard-related stuff for MANY years now. :)

I did a little bit of catching-up by looking at recent issues here and updates to the Italian "CNS" specs. I noticed that the file system spec was updated in 2016 to say that "2048-bit keys are hard, there's no widespread standard solution, so vendors are allowed to disregard this documentation as long as they supply some middleware that works", which is somewhat disappointing.

The APDU logs are deleted where the certificate is read (even in the read with OpenSC), so I can't tell for sure, but I have 2 hypotheses:
1.  the cert is encoded in an uncommon / non-standard way and the driver can't read it for that reason;
2. the vendor dumped a dummy file in the `EF.C_Carta` location, because the card expects another set of commands to use it.

@nestukh, could you check if the APDUs there contain something that looks like a cert or just zeros?

Thanks!

**Comment by nestukh** [comment]
@puccia in the logs, `39: C_Digest` has non zero data, then after `112 CKR_MECHANISM_INVALID` all `C_Digest` functions return the same digit-only random multiline hex data. the rest of the `C_Digest*` and `C_Generate*` fields are alphanumeric random hex data

@resoli sorry, my bad, here it is
```
Using slot 0 with a present token (0x0)
Logging in to "Carta Nazionale dei Servizi".
Please enter User PIN: 
C_SeedRandom() and C_GenerateRandom():
  seems to be OK
Digests:
  all 4 digest functions seem to work
  SHA-1: OK
  SHA256: OK
  ERR: C_Digest() didn't return CKR_BUFFER_TOO_SMALL but CKR_GENERAL_ERROR (0x5)
  ERR: C_Digest() didn't return CKR_OK for a NULL output buffer, but CKR_OPERATION_NOT_INITIALIZED (0x91)
  ERR: digest operation ended prematurely
Signatures (currently only for RSA)
  testing key 0 (CNS1) 
  ERR: C_SignUpdate failed: CKR_FUNCTION_FAILED (0x6)
  ERR: C_Sign() didn't return CKR_BUFFER_TOO_SMALL but CKR_GENERAL_ERROR (0x5)
  ERR: C_Sign() didn't return CKR_OK for a NULL output buf, but CKR_OPERATION_NOT_INITIALIZED (0x91)
error: PKCS11 function C_Sign failed: rv = CKR_GENERAL_ERROR (0x5)
Aborting.
```

**Comment by puccia** [comment]
@nestukh I am referring to the logs that were generated by `OPENSC_DEBUG=3 $HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /home/$USER/.bin/TS-CNS/static/lib/opensc-pkcs11.so` (i.e. https://gist.github.com/nestukh/29ddd4e55f63bcb207e8f158d41a69d3).

Thanks!

**Comment by nestukh** [comment]
@puccia sorry, 
after line 1475 (`Incoming APDU (257 bytes):`) there is data (I can read my Fiscal Code and name in it), that's the part I've omitted from the logs; the rest of big hex dumps are _just zeros_, yes

**Comment by puccia** [comment]
Actually, apologies — reading those logs (lines 1543–2094) I can confirm that they are zeros. So the driver thinks there's a certificate there, but it isn't (because the new specs allow that).

I think that the next step is finding out where and how the certs are read, perhaps by dumping the APDUs from a successful auth session.

**Comment by nestukh** [comment]
How do I log that? Using the official driver the output of the command ```$HOME/.bin/TS-CNS/static/bin/pkcs11-tool --login --test --module /usr/lib/bit4id/libbit4opki.so``` doesn't differ with or without `OPENSC_DEBUG=3` (as written in the opening message). I can't login with the `opensc-pkcs11.so` module, only with `libbit4opki.so`

**Comment by puccia** [comment]
Maybe there is a quicker way to do it, but here's what I used to do: run `pcscd` in debug mode with apdu logging (https://www.tutorialspoint.com/unix_commands/pcscd.htm, so options `-a -d`, and `-f` if you want to just keep it in the foreground instead of running it as a daemon).

**Comment by nestukh** [comment]
So on Debian, I've run `sudo /etc/init.d/pcscd stop; sudo /usr/sbin/pcscd -a -d -f`, and the resulting log for a successful login with `$HOME/.bin/TS-CNS/static/bin/pkcs11-tool` via `libbit4opki.so` is [here](https://gist.github.com/nestukh/f9724b216161dd814b866e4089728ef3) (I have `Ctrl+c`-ed at the end in order to stop it)

**Comment by resoli** [comment]
> I haven't worked on any smartcard-related stuff for MANY years now. :)

Hello @puccia, nice to here from you!

> I did a little bit of catching-up by looking at recent issues here and updates to the Italian "CNS" specs. I noticed that the file system spec was updated in 2016 to say that "2048-bit keys are hard, there's no widespread standard solution, so vendors are allowed to disregard this documentation as long as they supply some middleware that works", which is somewhat disappointing.

Nice understatement ;-). The document, for who reads italian, is [here](https://www.agid.gov.it/sites/default/files/repository_files/documentazione/filesystemcns_20160610.pdf) (page 21).  

In particular:

> The use of the 2048-bit authentication key on the CNS can be done in a chip-specific manner; 
> - The objects necessary for the management of such a key and its certificate can be placed in the part of the CNS file system not regulated by this specification (e.g.: DF_DS, a dedicated DF or an applet).  In this case, the CNS file system objects EF_KeyPub, EF_C_Carta, BSO.KPri are replaced by the homologous objects in the above mentioned DFs, ensuring equivalent  
> security levels; 
> - The use of the keys must be managed by the middleware made available by the supplier, possibly also through APDU commands not defined in the CNS functional specifications; 
> - The middleware must be made available free of charge for the development of applications that interact with the smartcards and to smartcard holders together with a procedure that allows simple installation.

This means that any constraints on the vendor are gone now. We can say hello to the stability and level of interoperability that OpenSC users have enjoyed for many years in their CNS usage. 

On the same page, we can read, as  a consequence of 2048 bit keys: "This impacts not only the handling of the single command APDU, for which there are ISO standard extensions such as command chaining or extended APDUs, but also the data structures used to store security quantities, such as the TLV.  The solutions  n the market as of June 2016, in fact, use heterogeneous systems to overcome these limitations". Which accounts for problems reported with ST 2020 CNS cards in https://github.com/OpenSC/OpenSC/issues/2370 .

**Comment by 3v1n0** [comment]
Hey, I've hacked a bit on the new Card to figure out the new paths needed to get it working with OpenSC.

For future reference, the best way to figure stuff out has been to use Wireshark (decoding pakets as USB CCID first and then as ISO 7816), while doing various operations with the proprietary driver (like `pkcs11-tool --module /usr/lib/bit4id/libbit4xpki.so -T`).

There's a missing bit though: the random number generator, as the proprietary driver does it in software... :vomiting_face:.

Stuff like signature can be tested atomically (i.e. without using `-t`) with stuff like:

```bash
echo 'Ciao!' > ciao
openssl dgst -sha256 -binary ciao > ciao.sha256
pkcs11-tool --signature-format openssl --sign -o ciao.signed --mechanism RSA-PKCS-PSS --hash-algorithm SHA256 --input ciao.sha256
# otherwise
pkcs15-crypt -s -i ciao.sha256 -o ciao.signed -f openssl --sha-256 --pkcs1
pkcs11-tool -r --label CNS1 --type pubkey | openssl rsa -pubin -inform DER -outform PEM > pubkey.pem
openssl dgst -sha256 -verify pubkey.pem -signature ciao.signed ciao
```

I've done some tests, but let me know if you find troubles or can help in improving the card detection (I couldn't find another way than rely on manufacturer) or random number generation.

**Comment by 3v1n0** [comment]
Ah, I also added support for the NFC ATR value, so that the card can now used also with a contactless reader.

Worth to mention, that ANY Android phone can be easily used as a (wireless/remote) NFC reader, when used with [vsmartcard](https://frankmorgner.github.io/vsmartcard/remote-reader/README.html) (also available in debian as `vsmartcard-vpcd`).

**Comment by nestukh** [comment]
Sorry if I'm late to reply.
vsmartcard is very cool, but my phone hasn't got NFC, I'll try it in the future.
I'm waiting for the commit to be accepted into master, then I recompile then I'll post results.
hugs

**Comment by resoli** [comment]
@3v1n0 really great! I'm in contact with peoples trying to use OpenSC on their new ACe 2021 (Idemia) CNS. May be they are willing to test. Very clever also your suggestion about vsmartcard.
Someone is planning to implement support for CIEv3 in OpenSC, are you interested? I will contact you in private, if so.

**Comment by 3v1n0** [comment]
> Someone is planning to implement support for CIEv3 in OpenSC, are you interested? I will contact you in private, if so.

Oh  indeed I am, I was already hacking on it in these days... But I'm still at the very early stages (just some basic initialization), as I'm studying the stuff and from what I see the quality of the official driver is just.... :vomiting_face:.
Feel free to ping me on IRC.

**Comment by resoli** [comment]
> Feel free to ping me on IRC.
where? Trevinho sul canale #ubuntu-unity server irc.freenode.net ?

**Comment by ilarioq** [comment]
Excuse my English made in Deepl!
I was able to authenticate with an ACE2021 card.

**Comment by 3v1n0** [comment]
> > Feel free to ping me on IRC.
> 
> where? Trevinho sul canale #ubuntu-unity server irc.freenode.net ?

Yeah, I'm `Trevinho` in any server, better would be libera chat these days or OFTC (or even gimpnet).

**Comment by resoli** [comment]
@3v1n0 I am trying to contact you via email (the one in your github profile. My msg is very likely  marked as spam)

**Comment by nestukh** [comment]
This afternoon my mind isn't overwhelmed with work-related stuff and I've checked out your pull request, it works as you said. **Yay!**

for anyone in need, until the devs merge it into the origin branch, just do
```
[...]
git clone https://github.com/OpenSC/OpenSC
cd OpenSC
git fetch origin pull/2483/head && git checkout FETCH_HEAD
[...]
```
and then continue with bootstrapping and compiling as suggested in a comment above.

As an additional note: maybe in the future the main tree won't be compatible with the commits (bad luck is always around the corner), but right now it's working, so just in case you need to revert changes as it is today, add another line in the middle:
```
[...]
git clone https://github.com/OpenSC/OpenSC
cd OpenSC
git checkout $(git rev-list -n 1 --before="2022-08-15 15:15" $(git rev-parse --abbrev-ref HEAD))
git fetch origin pull/2483/head && git checkout FETCH_HEAD
[...]
```
That's it!
Thanks guys for the incredible help.
